File: Chapter03/Activity3_03/Activity3_03.scala
package Activity3_03

import Utilities02.HelperScala.{extractRawRecords, parseRawWarc}
import Utilities02.WarcRecord
import org.apache.commons.io.IOUtils
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import org.apache.tika.langdetect.OptimaizeLangDetector
import org.apache.tika.metadata.Metadata
import org.apache.tika.parser.ParseContext
import org.apache.tika.parser.html.HtmlParser
import org.apache.tika.sax.BodyContentHandler

object Activity3_03 {

  def tagRecords(partition: Iterator[WarcRecord]): Iterator[(String, String, Float)] = {
    val textHandler = new BodyContentHandler(-1) // -1 skips document length limit
    val metadata = new Metadata()
    val parser = new HtmlParser()
    val context = new ParseContext()
    val languageIdentifier = new OptimaizeLangDetector()
    languageIdentifier.loadModels()

    partition.flatMap(record => {
      val html = record.htmlSource
      if (html.nonEmpty) {
        parser.parse(IOUtils.toInputStream(html, "UTF-8"), textHandler, metadata, context)
        val extractedText = textHandler.toString.trim.replaceAll("\\s+", " ")
        val detected = languageIdentifier.detect(extractedText)
        println(record.targetURI, detected.getLanguage, detected.getRawScore)
        Some(record.targetURI, detected.getLanguage, detected.getRawScore)
      }
      else
        None
    })
  }

  def main(args: Array[String]): Unit = {
    implicit val session = SparkSession.builder
      .appName("Improved Crawl Tagger")
      .getOrCreate()
    val input = args(0)
    val outputDir = args(1)

    val warcRecords: RDD[WarcRecord] = extractRawRecords(input)
      .flatMap(parseRawWarc)
      .filter(_.warcType == "response")

    val taggedTexts = warcRecords.mapPartitions(tagRecords)
    taggedTexts.saveAsTextFile(outputDir)

  }
}

File: Chapter03/Activity3_03/Activity3_03.py
from sys import argv
from bs4 import BeautifulSoup
from re import sub
from string import printable
import pycld2
from pyspark.sql import SparkSession
from Chapter02.utilities02_py.helper_python import extract_raw_records, parse_raw_warc


def tag_records(partition):
    for warc_record in partition:
        parser = BeautifulSoup(warc_record.html_source, 'html.parser')
        plaintext = ' '.join(parser.stripped_strings)
        plaintext_stripped = sub('\\s+', ' ', plaintext)
        if plaintext_stripped is None or plaintext_stripped == '':
            yield ()  # empty tuple
        else:
            cleaned_text = ''.join(x for x in plaintext_stripped if x in printable)
            _, _, details = pycld2.detect(cleaned_text)
            (languageName, languageCode, percent, score) = details[0]
            yield warc_record.target_uri, languageCode, str(score)


if __name__ == "__main__":
    session: SparkSession = SparkSession.builder \
        .appName('Improved Crawl Tagger') \
        .getOrCreate()
    input_file = argv[1]
    output_dir = argv[2]
    warc_records = extract_raw_records(input_file, session) \
        .flatMap(lambda record: parse_raw_warc(record)) \
        .filter(lambda record: record.warc_type == 'response')

    tagged_texts_rdd = warc_records \
        .mapPartitions(tag_records) \
        .filter(lambda record: record != ())

    tagged_texts_rdd.saveAsTextFile(output_dir)
